resolver/dns/dns_resolver.go

```diff
@@ -146,6 +151,7 @@ func (b *dnsBuilder) Build(target resolver.Target, cc resolver.ClientConn, opts
 	}
 
 	d.wg.Add(1)
+	go d.rateLimiter()
```

Comment:
Can you find an approach that doesn't require a new goroutine?  E.g. we could keep the time of the last re-resolve in the watcher, and then, when a re-resolve request comes in, sleep until that time plus `minDNSResRate`.

Reply:
Since we need to take care of context expiration as well, we cannot simple sleep. So, I would have to do something like this:

- Add a `lastResolved time.Time` in `dnsResolver` initialized to the zero value of time.Time
- In `watcher()`, after the `select`, we would call `waitBeforeResolution()`, and if returns false, we would have to return from there. Else we reset lastResolved to time.Now() and proceed with resolution.
```
- func (d *dnsResolver) waitBeforeResolution() bool {
    now := time.Now()
    waitFor := time.Duration(0)
    if next := d.lastResolved.Add(minDNSResRate); next.After(now) {
      waitFor = next.Sub(now)
    }

    tr := time.NewTimer(next)
    select {
    case <-d.ctx.Done():
      if !tr.Stop() {
        <-tr.C
      }
      return false
    case <-tr.C:
      return true
    }
}
```
Is this what you were suggesting?
And why do you not prefer using a goroutine? 
Thanks.

Reply:
I think you can make it a little simpler, even, avoiding the struct field and some other minor tweaks:

```go
func watcher() {
  for {
    select { ... } // as it is today
    // resolve
    t := time.NewTimer(minDNSResRate)
    select {
	case <-t.C:
    case <-d.ctx.Done():
      t.Stop()  // Don't need to drain t.C since we don't check it
      return
    }
  }
}
```

goroutines have an overhead, both in resources and code complexity, so it's best to avoid them when possible.

Reply:
Done. This is nice, much simpler. Thanks.

Reply:
Btw, how do you get syntax highlighting in your code blocks?